llvm/include/llvm/BinaryFormat/DXContainer.h

```diff
@@ -546,13 +560,75 @@ struct ProgramSignatureElement {
 static_assert(sizeof(ProgramSignatureElement) == 32,
               "ProgramSignatureElement is misaligned");
 
+struct RootConstants {
+  uint32_t ShaderRegister = 0;
+  uint32_t RegisterSpace = 0;
+  uint32_t Num32BitValues = 0;
+
+  RootConstants() = default;
+
+  void swapBytes() {
+    sys::swapByteOrder(ShaderRegister);
+    sys::swapByteOrder(RegisterSpace);
+    sys::swapByteOrder(Num32BitValues);
+  }
+};
+
+struct RootParameter {
+  dxbc::RootParameterType ParameterType;
+  union {
+    dxbc::RootConstants Constants;
+  };
+  dxbc::ShaderVisibilityFlag ShaderVisibility;
+
+  RootParameter() {
+    Constants = RootConstants();
+    ParameterType = dxbc::RootParameterType::Empty;
+    ShaderVisibility = dxbc::ShaderVisibilityFlag::Empty;
+  }
+
+  void swapBytes() {
+
+    sys::swapByteOrder(ParameterType);
+    sys::swapByteOrder(ShaderVisibility);
+    switch (ParameterType) {
+    case RootParameterType::Constants32Bit:
+      Constants.swapBytes();
+      break;
+    case RootParameterType::Empty:
+      llvm_unreachable("invalid value for ParameterType");
+      break;
+    }
+  }
+};
+
+struct RootSignatureHeader {
+  uint32_t Version = 2;
+  uint32_t Flags = 0;
+
+  void swapBytes() {
+    sys::swapByteOrder(Version);
+    sys::swapByteOrder(Flags);
+  }
+};
+
 struct RootSignatureValidations {
 
   static bool isValidRootFlag(uint32_t Flags) { return (Flags & ~0xfff) == 0; }
 
   static bool isValidVersion(uint32_t Version) {
     return (Version == 1 || Version == 2);
   }
+
+  static bool isValidParameterType(dxbc::RootParameterType Flag) {
+    // RootParameterType::Empty is the higest value in the enum.
+    return Flag < dxbc::RootParameterType::Empty;
```

Comment:
I wonder if the sets of flags here are small enough that we could have a positive-check here that explicitly checks for ones that are known to be valid?